### PR TITLE
Enable gauntlet training

### PR DESCRIPTION
**Pull request by bmosaicml**
This PR enables us to run the gauntlet during training. It also enables you to set a subset of batches to run eval over in order to speed up ICL during training.

It also changes eval to no longer use the InMemoryLogger and instead pull the metrics from the State.

Eval test run:  `all-eval-F2oe1D` 
Training test run: `test-1b-C5QG5k` `test-1b-c7sqCh `

https://wandb.ai/mosaic-ml/gauntlet/runs/rrbn0qbb?workspace=user-jemdohmann

image

Eval results still good:

```
| model_name               |   average |   world_knowledge |   commonsense_reasoning |   language_understanding |   symbolic_problem_solving |   reading_comprehension |
|:-------------------------|----------:|------------------:|------------------------:|-------------------------:|---------------------------:|------------------------:|
| mosaicml/mpt-7b-instruct |  0.354255 |          0.398764 |                0.415097 |                 0.371509 |                   0.171216 |                0.414691 |
```

### CODE REVIEW

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (abhi-databricks) [inline llmfoundry/callbacks/model_gauntlet_callback.py:130]: @bmosaicml what's the reason we need to grep the logs from the InMemoryLogger rather than read the eval metrics out of the state directly?

File: scripts/eval/yamls/mpt_eval.yaml
Comment (dakinggg) [inline scripts/eval/yamls/mpt_eval.yaml:27]: this indentation seems wrong?

File: scripts/eval/yamls/hf_eval.yaml
Comment (dakinggg) [inline scripts/eval/yamls/hf_eval.yaml:40]: should this be commented out?

File: scripts/eval/yamls/hf_eval.yaml
Comment (dakinggg) [inline scripts/eval/yamls/hf_eval.yaml:46]: is this change intentional?

File: scripts/eval/eval.py
Comment (dakinggg) [inline scripts/eval/eval.py:152]: why are we using markdown results? This seems like an odd choice for outputting structured data

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (dakinggg) [inline llmfoundry/callbacks/model_gauntlet_callback.py:108]: when would this condition happen?

File: scripts/eval/eval.py
Comment (dakinggg) [inline scripts/eval/eval.py:143]: shouldn't the `eval_after_all` just write some data to the `model_gauntlet_callback` (or to a file if you prefer)? Then you don't need to call the event explicitly, as it should just run on the event as intended after eval.

File: llmfoundry/utils/builders.py
Comment (dakinggg) [inline llmfoundry/utils/builders.py:184]: this shouldn't be type `Any`, it should be a `DictConfig`

Comment (dakinggg) [reply llmfoundry/utils/builders.py:184]: you probably just need to move the assert below above the call to this function

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (hanlint) [inline llmfoundry/callbacks/model_gauntlet_callback.py:28]: `Exception` should almost never be used. In thise case, you should throw a `ValueError`.

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (hanlint) [inline llmfoundry/callbacks/model_gauntlet_callback.py:110]: was the index change from `[0]` -> `[-1]` intentional here?

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (hanlint) [reply llmfoundry/callbacks/model_gauntlet_callback.py:130]: +1 this is not very robust and strongly prefer a better way here.

File: mcli/mcli-hf-eval.yaml
Comment (hanlint) [inline mcli/mcli-hf-eval.yaml:4]: We need a tagged branch here, can't be tip of dev as that is too unstable.

File: mcli/mcli-1b-gauntlet.yaml
Comment (hanlint) [inline mcli/mcli-1b-gauntlet.yaml:17]: Please remove specific internal cluster names

File: mcli/mcli-1b-gauntlet.yaml
Comment (hanlint) [inline mcli/mcli-1b-gauntlet.yaml:13]: our yaml examples should not have specific mosaicml setups.

File: mcli/mcli-1b-gauntlet.yaml
Comment (hanlint) [inline mcli/mcli-1b-gauntlet.yaml:101]: please remove hardcoded values

File: mcli/mcli-1b-gauntlet.yaml
Comment (hanlint) [inline mcli/mcli-1b-gauntlet.yaml:183]: remove

File: mcli/mcli-1b-gauntlet.yaml
Comment (hanlint) [inline mcli/mcli-1b-gauntlet.yaml:188]: remove

File: mcli/mcli-hf-eval.yaml
Comment (hanlint) [inline mcli/mcli-hf-eval.yaml:50]: why are these needed?

File: scripts/eval/yamls/mpt_eval.yaml
Comment (hanlint) [inline scripts/eval/yamls/mpt_eval.yaml:1]: why this change?

File: scripts/eval/yamls/mpt_eval.yaml
Comment (hanlint) [inline scripts/eval/yamls/mpt_eval.yaml:35]: Why are these changed?

File: scripts/train/train.py
Comment (hanlint) [inline scripts/train/train.py:259]: Shouldn't there be a different type check here?

File: scripts/train/train.py
Comment (hanlint) [inline scripts/train/train.py:289]: we should never silently add an unwanted callback. suggest we find a different way to pass data, rather than rely on a logger destination.

**Review by hanlint — CHANGES_REQUESTED** [review]
Thanks for putting this PR together, left a few comments:
* the new yaml files have unexplained changes (batch size, fsdp config), and also have several hardcoded internal OCI bucket names and cluster names. Please remove the hardcoded values prior to pushing PRs.
* can we avoid silently adding in an InMemoryLogger, purely for the purposes of passing around data? Seems like there's a deeper design issue if we have to resort to this approach.

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (bmosaicml) [reply llmfoundry/callbacks/model_gauntlet_callback.py:110]: yes, since logging events to the same key append to a list, in order to retrieve the most recent logging we need to use index -1

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (eracah) [reply llmfoundry/callbacks/model_gauntlet_callback.py:130]: I think you should use `state.eval_metric_values` here. It should contain a dictionary of all recently computed metrics, keyed by metric name. The one issue is it is not keyed by dataloader name, so if there are multiple dataloaders with the same metric there will be a collision, but we can patch in that change to enable this PR

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (eracah) [reply llmfoundry/callbacks/model_gauntlet_callback.py:110]: `state.eval_metric_values` will contain the most recent eval metrics

File: scripts/train/train.py
Comment (eracah) [reply scripts/train/train.py:289]: I thiiink `state.eval_metric_values` should work here

**Review by eracah — COMMENTED** [review]
Might be worth using state to get the metrics instead of `inmemorylogger`

File: mcli/mcli-hf-eval.yaml
Comment (bmosaicml) [reply mcli/mcli-hf-eval.yaml:50]: Brandon found that it made FSDP + inference faster to use these flags

File: scripts/eval/yamls/mpt_eval.yaml
Comment (bmosaicml) [reply scripts/eval/yamls/mpt_eval.yaml:1]: this is the standard length used for our gauntlet evals

File: scripts/eval/yamls/hf_eval.yaml
Comment (bmosaicml) [reply scripts/eval/yamls/hf_eval.yaml:46]: yeah I want all the configs to reflect the standard gauntlet usage

File: scripts/eval/eval.py
Comment (bmosaicml) [reply scripts/eval/eval.py]: I'm open to other suggestions. What did you have in mind?

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (hanlint) [inline llmfoundry/callbacks/model_gauntlet_callback.py:30]: This is API breaking -- is this an absolutely neccessary change, and if so, can we have backwards compatibility with a deprecation warning?

File: tests/test_training.py
Comment (hanlint) [inline tests/test_training.py:114]: tests need some sort of assert or check

File: mcli/mcli-hf-eval.yaml
Comment (hanlint) [reply mcli/mcli-hf-eval.yaml:4]: I don't think this was resolved?

Comment (bmosaicml) [reply mcli/mcli-hf-eval.yaml:4]: Isn't v.0.2.0 stable? We can also just comment it out entirely and let users put whatever tag/branch they want

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (bmosaicml) [reply llmfoundry/callbacks/model_gauntlet_callback.py:30]: It's not strictly necessary, so I can change it back!

File: tests/test_training.py
Comment (bmosaicml) [reply tests/test_training.py:114]: I think ideally we would check that the ModelGauntlet was constructed and its eval_after_all method was called but I am not if it's possible to Mock that if the object is constructed within Main. Fwiw, the other unit test in this method has no asserts and simply checks whether the code fails

Comment (dakinggg) [reply tests/test_training.py:114]: Yeah, these tests are really "does it crash". For the gauntlet, there should separately be unit tests testing the basic functionality.

Comment (bmosaicml) [reply tests/test_training.py:117]: I think it is important to test the full workflow of running train and seeing if it logs properly. 

I am trying to unit test this by using console logger but I think the definition of console logger's fit_end and eval_end methods preclude the gauntlet callback from working properly. This is because the gauntlet callback runs at eval_after_all. As it currently works, the console logger runs at eval_end and at fit end, but anything logged between eval_end and fit_end (e.g. during eval_after_all) will just sit in purgatory and not get flushed.

Lmk how you think I should proceed.

File: mcli/mcli-hf-eval.yaml
Comment (hanlint) [reply mcli/mcli-hf-eval.yaml:4]: Yes `v0.2.0` is stable, but this diff comments it out, which means that this yaml defaults to using the tip of `dev`. I think this is OK, but we should update this to a version upon the next release. Defer to @dakinggg here.

Comment (dakinggg) [reply mcli/mcli-hf-eval.yaml:4]: My preference is to have the MCLI yamls remain pinned to releases and then update them when we do a release. If we really _need_ to break this so we can give the yaml to a customer, it is ok. But the yaml should pin to a specific commit that works (e.g. https://github.com/mosaicml/llm-foundry/blob/667006889b5aab99fbecf1f927704fa7645e1799/mcli/mcli-llama2-finetune.yaml#L1-L6 for llama2)

File: tests/test_training.py
Comment (dakinggg) [reply tests/test_training.py:126]: I guess `ConsoleLogger` needs to flush on `eval_after_all`? Can you make a JIRA for that and assign it to Evan please?

In the meantime...can you test using `InMemoryLogger`?

Comment (bmosaicml) [reply tests/test_training.py:126]: I'm not sure that InMemoryLogger will work. How will I access the logger's data from inside the test?

File: mcli/mcli-hf-eval.yaml
Comment (bmosaicml) [reply mcli/mcli-hf-eval.yaml:4]: Should I add v.0.2.0 back in?

File: tests/test_training.py
Comment (dakinggg) [reply tests/test_training.py:125]: I'm assuming in the test you are talking about you want to test calling `main`. Assuming that is true, you can return the `Trainer` from main, and then use the returned object to get whatever other objects you need.

Comment (bmosaicml) [reply tests/test_training.py:125]: ok I will modify main to return trainer...good call

**Comment by abhi-mosaic** [comment]
Sorry if this is a silly question, but is it meant to be `gauntlet` or `model_gauntlet`? I feel like the former is a bit more accurate? Or even `eval_gauntlet`, but `model_gauntlet` sounds a bit strange for a callback that averages eval metrics.

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (hanlint) [inline llmfoundry/callbacks/model_gauntlet_callback.py:46]: Since this is an API change, can we add in a check below, e.g.
```
if isinstance(logger_keys, dict):
   raise ValueError('logger_keys now requires a list type as input, not a dict') 
```

The code path now seems to support both, but down the road, people might have silent failures if we start depending on the type being a list instead of a dict.

**Review by hanlint — APPROVED** [review]
LGTM, with one minor suggestion. But please get an approval from another reviewer before merging.

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (bmosaicml) [reply llmfoundry/callbacks/model_gauntlet_callback.py:46]: done!

**Comment by bmosaicml** [comment]
> Sorry if this is a silly question, but is it meant to be `gauntlet` or `model_gauntlet`? I feel like the former is a bit more accurate? Or even `eval_gauntlet`, but `model_gauntlet` sounds a bit strange for a callback that averages eval metrics.

I've been calling it model gauntlet since we run models through it haha. We could call it just gauntlet or eval gauntlet if you prefer? just lmk

**Review by abhi-databricks — APPROVED** [review]
Excited to start using this in our training runs! Thanks @bmosaicml for getting this over the finish line.

As discussed offline, I think we should change the name from `model_gauntlet` to `gauntlet` or `eval_gauntlet`, up to you. Approving to unblock but will leave it to you to rename and merge.

File: llmfoundry/callbacks/model_gauntlet_callback.py
Comment (dakinggg) [inline llmfoundry/callbacks/model_gauntlet_callback.py:100]:
```suggestion
            # starting at index 1 skips the "metric" part of the key which is superfluous
```

File: scripts/__init__.py
Comment (dakinggg) [inline scripts/__init__.py:1]: why do we need to make scripts a module? Would be better to not, and anything that needs to be importable move into the llmfoundry folder. If its just needed for tests, you can just append to the path to make things importable.

File: scripts/__init__.py
Comment (dakinggg) [inline scripts/__init__.py:10]: I don't think the version info belongs here?

File: scripts/train/__init__.py
Comment (dakinggg) [inline scripts/train/__init__.py:1]: same q

File: tests/test_eval.py
Comment (dakinggg) [inline tests/test_eval.py:19]: why do we need this?

**Review by dakinggg — COMMENTED** [review]
The tests that you have added are really slow, nearly doubling the time it takes CI to run

```
196.68s call     tests/test_training.py::test_train_gauntlet
161.70s call     tests/test_training.py::test_train[cpu]
120.18s call     tests/test_model_gauntlet.py::test_gauntlet_callback[True-False]
120.00s call     tests/test_model_gauntlet.py::test_gauntlet_callback[False-True]
119.20s call     tests/test_model_gauntlet.py::test_gauntlet_callback[True-True]
118.90s call     tests/test_model_gauntlet.py::test_gauntlet_callback[False-False]
```

Can you please reduce the intensity of the tests? Probably run less eval tasks or subset batches or something?

File: tests/test_eval.py
Comment (bmosaicml) [reply tests/test_eval.py:19]: tasks.yaml contains a bunch of dataset paths that are relative to the scripts directory

**Comment by bmosaicml** [comment]
> The tests that you have added are really slow, nearly doubling the time it takes CI to run
> 
> ```
> 196.68s call     tests/test_training.py::test_train_gauntlet
> 161.70s call     tests/test_training.py::test_train[cpu]
> 120.18s call     tests/test_model_gauntlet.py::test_gauntlet_callback[True-False]
> 120.00s call     tests/test_model_gauntlet.py::test_gauntlet_callback[False-True]
> 119.20s call     tests/test_model_gauntlet.py::test_gauntlet_callback[True-True]
> 118.90s call     tests/test_model_gauntlet.py::test_gauntlet_callback[False-False]
> ```
> 
> Can you please reduce the intensity of the tests? Probably run less eval tasks or subset batches or something?

Not sure `test_train_gauntlet` can be sped up anymore as it's now about the same speed as test_train. Should I just cut it? Or make it GPU only?

File: mcli/mcli-1b-eval.yaml
Comment (dakinggg) [inline mcli/mcli-1b-eval.yaml:25]: does this yaml run using foundry release 0.2.0. I'm assuming not. We can't change yamls that are pinned to a foundry release out of sync with new releases.

File: mcli/mcli-hf-eval.yaml
Comment (dakinggg) [inline mcli/mcli-hf-eval.yaml:54]: same version compatibility comment

File: scripts/eval/eval.py
Comment (dakinggg) [inline scripts/eval/eval.py:299]:
```suggestion
        # dl_name is either 2-tuple (benchmark_name, num_fewshot)
```

File: scripts/eval/yamls/tasks_lite.yaml
Comment (dakinggg) [inline scripts/eval/yamls/tasks_lite.yaml:1]: Can we please not change the name of this file? It will break many yamls.

File: scripts/train/train.py
Comment (dakinggg) [inline scripts/train/train.py:412]: this doesn't seem right. why is this code changed?

File: scripts/train/train.py
Comment (dakinggg) [inline scripts/train/train.py:253]: In order to not break existing yamls, please check for both `model_gauntlet` and `eval_gauntlet`, and emit a deprecation warning if `model_gauntlet` is used.

File: scripts/eval/eval.py
Comment (dakinggg) [inline scripts/eval/eval.py:49]: This change doesn't seem like it should be part of this pr?

File: scripts/eval/eval.py
Comment (dakinggg) [inline scripts/eval/eval.py:168]: same comment as train script. need to not break existing yamls, so look for both old key and new key.

**Review by dakinggg — CHANGES_REQUESTED** [review]
see remaining comments around backwards compatibility and a couple code changes that I think were a bad merge.

File: llmfoundry/callbacks/eval_gauntlet_callback.py
Comment (dakinggg) [inline llmfoundry/callbacks/eval_gauntlet_callback.py:22]: Ah I didn't notice initially that the file name changed too. So, this should actually be in the original file name `model_gauntlet_callback.py`

File: mcli/mcli-1b-eval.yaml
Comment (dakinggg) [inline mcli/mcli-1b-eval.yaml:4]: This still needs to be pinned, so if this yaml is updated to only work with main, we need to repin after this PR goes in.

File: scripts/train/train.py
Comment (dakinggg) [inline scripts/train/train.py:505]: Can you put the comment back?

File: mcli/mcli-1b-eval.yaml
Comment (bmosaicml) [reply mcli/mcli-1b-eval.yaml:4]: what should I pin it to?

**Review by dakinggg — APPROVED** [review]

**Pull request merged**
